File: dsa-dart/test/stack/stack_test.dart
import 'package:dsa_dart/dsa_dart.dart';
import 'package:dsa_dart/stack/stack.dart';
import 'package:test/test.dart';

void main() {
  test('calculate', () {
    expect(calculate(), 42);
  });

  test('Stack - Push and Pop', () {
    Stack<int> stack = Stack<int>();
    stack.push(1);
    stack.push(2);
    stack.push(3);

    expect(stack.pop(), 3);
    expect(stack.pop(), 2);
    expect(stack.pop(), 1);
  });

  test('Stack - Peek', () {
    Stack<int> stack = Stack<int>();
    stack.push(1);
    stack.push(2);
    stack.push(3);

    expect(stack.peek(), 3);
    expect(stack.peek(), 3);
  });

  test('Stack - isEmpty', () {
    Stack<int> stack = Stack<int>();

    expect(stack.isEmpty, true);

    stack.push(1);

    expect(stack.isEmpty, false);
  });

  test('Stack - size', () {
    Stack<int> stack = Stack<int>();

    expect(stack.size, 0);

    stack.push(1);
    stack.push(2);
    stack.push(3);

    expect(stack.size, 3);
  });
}

File: dsa-ts/src/stack/stack.test.ts
import { describe, expect, test } from "vitest";

import { Stack } from "./stack";

describe("Stack", () => {
  test("should create an empty stack", () => {
    const stack = new Stack();
    expect(stack.size()).toBe(0);
    expect(stack.isEmpty()).toBe(true);
  });

  test("should push elements to the stack", () => {
    const stack = new Stack();
    stack.push(1);
    stack.push(2);
    expect(stack.size()).toBe(2);
    expect(stack.isEmpty()).toBe(false);
  });

  test("should pop elements from the stack", () => {
    const stack = new Stack();
    stack.push(1);
    stack.push(2);
    expect(stack.pop()).toBe(2);
    expect(stack.pop()).toBe(1);
    expect(stack.size()).toBe(0);
  });

  test("should peek the top element of the stack", () => {
    const stack = new Stack();
    stack.push(1);
    stack.push(2);
    expect(stack.peek()).toBe(2);
    stack.pop();
    expect(stack.peek()).toBe(1);
  });
});
